### PR TITLE
feat!: ping peers before adding to routing table

### PR DESCRIPTION
**Pull request by achingbrain**
Implements the [check-before-add](https://github.com/libp2p/go-libp2p-kad-dht/blob/master/optimizations.md#checking-before-adding) client optimisation to ping a peer before adding it to the routing table.

Adds a "new peer ping queue" to apply a concurrency limit to these pings, because it would be expected for old contacts to be less likely to be online so don't block adding new contacts to unrelated buckets if the connection to an old contact is timing out while being pinged before eviction.

BREAKING CHANGE: the routing ping options have been split into "old contact" and "new contact" and renamed according

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by maschad — COMMENTED** [review]
Look good overall, I left some comments as I do wonder if a consumer would be able to immediately understand what these new parameters are used for.

Also should we add a migration guide outlining these new params?

File: packages/kad-dht/src/index.ts [hunk 1]
```diff
@@ -400,15 +400,43 @@ export interface KadDHTInit {
...
+   */
+  pingNewContactTimeout?: Omit<AdaptiveTimeoutInit, 'metricsName' | 'metrics'>
+
+  /**
+   * How many peers to ping in parallel when deciding if they should be added to
+   * the routing table or not
+   *
+   * @default 10
+   */
+  pingNewContactConcurrency?: number
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/kad-dht/src/index.ts:432]: They are not the same. One says “added” and the other says “evicted”.

[the remark this answers is not in the document]
Comment (maschad) [reply packages/kad-dht/src/index.ts:432]: Right, oversight on my part

**Review by maschad — APPROVED** [review]
Nice work

**Pull request merged**